README.md
# Simple DateTime/Scheduler library

[![test](https://github.com/konrads/chrono-light/workflows/test/badge.svg)](https://github.com/konrads/chrono-light/actions/workflows/test.yml)

Provides `DateTime` and `Schedule` capabilities with no external library dependencies. Works in std/no std, optional support for [scale codec](https://github.com/paritytech/parity-scale-codec) encoding/decoding/type info.

Functionality provided:
* conversion from unixtime (mills from epoch) to `DateTime`, and vice versa
* finding next occurrence for a schedule comprising:
  * start `DateTime`
  * repeat frequency (multiples of year/month/week/day/hour/minute/second/millis)
  * optional end `DateTime`
* `scale` feature enabling `Encode`/`Decode`/`TypeInfo` support for `DateTime`/`Schedule` structs.

## Scope
This library works with `DateTime`s and `schedule`s within years of [1970, 4000].

Does not support timezones.

Overflow of months (>12), days (>28, >30, >31), hour (>23), minute/second (>59), millis (>999) is discouraged yet allowed, with excess added eg. 31 April ~= 1 May. Underflow of month/day (=0) causes panic. To avoid panic, validate hand crafted `DateTime` via `Calendar::validate()` or convert to unixtime via `Calendar::to_unixtime_opt()`.

Reasoning behind these restrictions is to keep the footprint reasonably compact and performance reasonably fast, and cater for real life scenarios.

### Typical usage
```rust
use chrono_light::prelude::*;

let c = Calendar::create();
let now_in_ms: u64 = 1650412800000;  // represents 20/04/2022 00:00:00:000
let schedule = Schedule {
    start: DateTime { year: 2020, month: 4, day: 30, hour: 0, minute: 0, second: 0, ms: 0 },
    items: vec![(Frequency::Year, 1)],
    end: Some(DateTime { year: 2025, month: 4, day: 30, hour: 0, minute: 0, second: 0, ms: 0 })
};
assert!(c.validate_schedule(&schedule).is_ok());
assert_eq!(Some(10*24*60*60*1000), c.next_occurrence_ms(&c.from_unixtime(now_in_ms), &schedule));  // triggers in 10 days
```

### Utility within a pallet

`Chrono-light` serves as a scheduling mechanism for [scheduler-pallet](https://github.com/konrads/pallet-scheduler-datetime).

CHANGELOG.md
# Changelog

## [0.1.5] 2022-07-06
- fixes for issues uncovered whilst playing with `Substrate` framework

## [0.1.4] 2022-06-26
- validation to utilize `Result`

## [0.1.3] 2022-06-24
- fixes to `no_std`

## [0.1.2] 2022-05-30
- fixes to scale codec

## [0.1.1] 2022-05-27
- added scale codec/info support

## [0.1.1] 2022-04-19
- initial implementation

src/types.rs
#[cfg(not(feature = "std"))]
use alloc::vec::Vec;
#[cfg(feature = "scale")]
use codec::{Decode, Encode};
#[cfg(feature = "scale")]
use scale_info::TypeInfo;

use super::constants::*;

/// DateTime representation from year to ms. Valid values are:
/// - year:   [1970, 4000]
/// - month:  [1, 12]
/// - day:    [1, 31] (depending on month, leap year)
/// - hour:   [0: 23]
/// - minute: [0, 59]
/// - second: [0, 59]
/// - ms:     [0, 999]
/// 
/// Note: other values will be accepted, but will be classified invalid by the calendar, and if used,
/// appropriate values will be added on top, eg. 32/01 -> 01/02.
#[derive(Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "scale", derive(Encode, Decode, TypeInfo))]
pub struct DateTime {
    // date
    pub year:   u16,
    pub month:  u8,
    pub day:    u8,

    // time
    pub hour:   u8,
    pub minute: u8,
    pub second: u8,
    pub ms:     u16,
}

impl DateTime {
    /// Calculates ms for the day
    pub fn to_day_unixtime(&self) -> u64 {
        self.day.checked_sub(1).expect("failed to calc day - 1") as u64 * MS_IN_DAY
            + self.hour as u64 * MS_IN_HOUR
            + self.minute as u64 * MS_IN_MIN
            + self.second as u64 * MS_IN_SEC
            + self.ms as u64
    }
}

/// Schedule, represented by a `start` `DateTime`, optional `end` `DateTime`, and multiple pairs of (`Frequency`, `multiplier`).
/// Next occurrence of trigger time is calculated by taking the earliest occurrence of `Frequency` * `multiplier`, from `start`, but before `end`.
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "scale", derive(Encode, Decode, TypeInfo))]
pub struct Schedule {
    pub start: DateTime,
    pub items: Vec<(Frequency, u32)>,  // frequency with multiplier
    pub end: Option<DateTime>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "scale", derive(Encode, Decode, TypeInfo))]
#[repr(u8)]
pub enum Frequency {
    Year,
    Month,
    Week,
    Day,
    Hour,
    Minute,
    Second,
    Ms,
}

impl Frequency {
    #[inline]
    pub fn to_ms(&self) -> u32 {
        match self {
            Frequency::Year   => 666_u32,
            Frequency::Month  => 999_u32,
            Frequency::Week   => 7 * MS_IN_DAY as u32,
            Frequency::Day    => MS_IN_DAY as u32,
            Frequency::Hour   => MS_IN_HOUR as u32,
            Frequency::Minute => MS_IN_MIN as u32,
            Frequency::Second => MS_IN_SEC as u32,
            Frequency::Ms     => 1_u32,
        }
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum ValidationError {
    /// `DateTime` not covered by this library, eg. 01/01/1000 00:00:00:000, 01/01/5000 00:00:00:000
    OutOfScope,
    /// Invalid `DateTime`, eg. 32/13/2000 66:66:66:6666, 29/02/2021 10:10:10:000 (non leap year)
    Invalid
}


src/tests_props.rs
use crate::constants::{EPOCH_YEAR, MS_IN_DAY, MS_IN_HOUR, MS_IN_SEC, MS_IN_MIN};
use super::prelude::*;

use chrono::{NaiveDateTime, TimeZone, Utc, Datelike, Timelike};
use quickcheck::QuickCheck;
use std::panic;

const PROPS_TESTS: u64 = 1000;  // 10 x the norm

/// For default quickcheck behaviour, with not panic hook initialization or test run settings, could also just do:
/// ```no_run
/// use quickcheck::quickcheck;
/// 
/// #[quickcheck]
/// fn validate_light_vs_chrono(...) {...}
/// ```
fn validate_light_vs_chrono(year: u16, month: u8, day: u8, hour: u8, minute: u8, second: u8, ms: u16) -> bool {
    let c = Calendar::create();

    // ensure values get some (but not too much) margin for error
    let year = year % (4000 - EPOCH_YEAR as u16) + EPOCH_YEAR as u16;
    let month = month % 15;
    let day = day % 35;
    let minute = minute % 65;
    let second = second % 65;
    let ms = ms % 1100;
    let dt_light = DateTime { year, month, day, hour, minute, second, ms};
    let dt_chrono_res = panic::catch_unwind(|| Utc.ymd(year as i32, month as u32, day as u32).and_hms_milli(hour as u32, minute as u32, second as u32, ms as u32));
    let dt_chrono_light_res = panic::catch_unwind(|| c.to_unixtime(&dt_light));
    let validation_result = c.validate_datetime(&dt_light);

    if validation_result == Err(ValidationError::Invalid) && dt_chrono_res.is_err() {
        true
    } else if validation_result == Ok(()) && dt_chrono_res.is_ok() && dt_chrono_light_res.is_ok() { // Note: dt_chrono_light_res may be ok when overflowing months/days, but never underflowing
        let dt_light_ms = c.to_unixtime(&dt_light);
        let dt_light2 = c.from_unixtime(dt_light_ms);
        let dt_chrono_ms = dt_chrono_res.unwrap().timestamp_millis() as u64;
        if dt_light_ms == dt_chrono_ms && dt_light == dt_light2 {
            true
        } else {
            println!("Failed on dt: {:?}, light ms: {:?}, chrono ms: {:?}", dt_light, dt_light_ms, dt_chrono_ms);
            false
        }
    } else {
        println!("Failed on dt: {:?}, validation_result: {:?}, dt_chrono_res: {:?}, dt_chrono_light_res: {:?}", dt_light, validation_result, dt_chrono_res, dt_chrono_light_res);
        false
    }
}

fn validate_scheduler_after_start(start_ms: u64, delta_ms: u64, freq: u8, freq_multiplier: u8) -> bool {
    let c = Calendar::create();

    fn get_exact_months(start_ms: u64, now_ms: u64, next_occurrence: u64) -> Option<u32> {
        let start = NaiveDateTime::from_timestamp(start_ms as i64 / 1000, ((start_ms % 1000) * 1_000_000) as u32);
        let end_ms = now_ms + next_occurrence;
        let end = NaiveDateTime::from_timestamp(end_ms as i64 / 1000, ((end_ms % 1000) * 1_000_000) as u32);
        if start.day() == end.day() && start.hour() == end.hour() && start.minute() == end.minute() && start.second() == end.second()
           && start.timestamp_subsec_millis() == end.timestamp_subsec_millis() {
            Some(((end.year() - start.year()) * 12 + end.month() as i32 - start.month() as i32) as u32)
        } else {
            None
        }
    }

    let freq = match freq % 8 {
        0 => Frequency::Year,
        1 => Frequency::Month,
        2 => Frequency::Week,
        3 => Frequency::Day,
        4 => Frequency::Hour,
        5 => Frequency::Minute,
        6 => Frequency::Second,
        _ => Frequency::Ms,
    };
    let freq_multiplier = freq_multiplier + 1;  // start at 1
    let mut start_ms = start_ms % 100*365*24*60*60*1000;
    let mut start = c.from_unixtime(start_ms);
    if freq == Frequency::Month {
        start.day = start.day.min(28);
        start_ms = c.to_unixtime(&start);
    }
    let delta_ms = delta_ms % 10*365*24*60*60*1000 + 1;
    let now_ms = start_ms + delta_ms;

    let now = c.from_unixtime(now_ms);
    let next_occurrence = c.next_occurrence_ms(&now, &Schedule { start: start.clone(), items: vec![(freq, freq_multiplier as u32)], end: None }).unwrap();

    match freq {
        Frequency::Year => {
            if let Some(months) = get_exact_months(start_ms, now_ms, next_occurrence) {
                months % (12 * freq_multiplier as u32) == 0 && next_occurrence <= freq_multiplier as u64 * MS_IN_DAY * 366
            } else {
                false
            }
        },
        Frequency::Month => {
            if let Some(months) = get_exact_months(start_ms, now_ms, next_occurrence) {
                months % freq_multiplier as u32 == 0 && next_occurrence <= freq_multiplier as u64 * MS_IN_DAY * 31
            } else {
                false
            }
        },
        _ => {
            let end_ms = now_ms + next_occurrence;
            let freq_ms = freq.to_ms() as u64 * freq_multiplier as u64;
            (end_ms - start_ms) % freq_ms == 0 && next_occurrence <= freq_ms
        }
    }
}


#[test]
fn test_validate_vs_chrono() {
    panic::set_hook(Box::new(|_info| { /* reduces the panic::catch_unwind() log noise */ }));
    QuickCheck::new().tests(PROPS_TESTS).max_tests(PROPS_TESTS * 100).quickcheck(validate_light_vs_chrono as fn(u16, u8, u8, u8, u8, u8, u16) -> bool)
}

#[test]
fn test_validate_scheduler_vs_chrono() {
    QuickCheck::new().tests(PROPS_TESTS).max_tests(PROPS_TESTS * 100).quickcheck(validate_scheduler_after_start as fn(u64, u64, u8, u8) -> bool)
}

#[test]
fn test_start_equals_now() {
    fn validate_start_equals_now(ts: u64, freq: u8) -> bool {
        let c = Calendar::create();
        let ts = ts % 60913560719000;
        let now = c.from_unixtime(ts);
        let freq = match freq % 8 {
            0 => Frequency::Year,
            1 => Frequency::Month,
            2 => Frequency::Week,
            3 => Frequency::Day,
            4 => Frequency::Hour,
            5 => Frequency::Minute,
            6 => Frequency::Second,
            _ => Frequency::Ms,
        };
        let freq_multiplier = freq.to_ms();

        fn is_trigger_valid(freq: Frequency, freq_multiplier: u32, x: u64) -> bool {
            match freq {
                Frequency::Year => x >= freq_multiplier as u64 * 365 * MS_IN_DAY && x <= freq_multiplier as u64 * 366 * MS_IN_DAY,  // estimate to avoid leap year calculation
                Frequency::Month => x >= freq_multiplier as u64 * 28 * MS_IN_DAY && x <= freq_multiplier as u64 * 31 * MS_IN_DAY,   // estimate to avoid month days calculation
                Frequency::Week => x == freq_multiplier as u64 * 7 * MS_IN_DAY,
                Frequency::Day => x == freq_multiplier as u64 * MS_IN_DAY,
                Frequency::Hour => x == freq_multiplier as u64 * MS_IN_HOUR,
                Frequency::Minute => x == freq_multiplier as u64 * MS_IN_MIN,
                Frequency::Second => x == freq_multiplier as u64 * MS_IN_SEC,
                Frequency::Ms => x == freq_multiplier as u64,
            }
        }

        let res = c.next_occurrence_ms(&now.clone(), &Schedule {
            start: now,
            items: vec![(freq, freq_multiplier)],
            end: None,
        });

        res.map_or(false, |x| is_trigger_valid(freq, freq_multiplier, x))
    }
    QuickCheck::new().tests(PROPS_TESTS / 10).max_tests(PROPS_TESTS * 10).quickcheck(validate_start_equals_now as fn(u64, u8) -> bool)
}

src/calendar.rs
#[cfg(not(feature = "std"))]
use alloc::vec::Vec;
use super::{
    constants::*,
    types::*,
    utils::*,
};

/// Representation of Calendar without timezones, has awareness of leap years, days in a given month for leap and non-leap years.
/// 
/// Typical usage
/// ```rust
/// use chrono_light::prelude::*;
/// let c = Calendar::create();
/// let now_in_ms: u64 = 1650412800000;  // represents 20/04/2022 00:00:00:000
/// let schedule = Schedule {
///     start: DateTime { year: 2020, month: 4, day: 30, hour: 0, minute: 0, second: 0, ms: 0 },
///     items: vec![(Frequency::Year, 1)],
///     end: Some(DateTime { year: 2025, month: 4, day: 30, hour: 0, minute: 0, second: 0, ms: 0 })
/// };
/// assert!(c.validate_schedule(&schedule).is_ok());
/// assert_eq!(Some(10*24*60*60*1000), c.next_occurrence_ms(&c.from_unixtime(now_in_ms), &schedule));  // triggers in 10 days
/// ```
/// 
/// Beware `c.to_unixtime()` may panic, use `c.validate()` and/or `c.to_unixtime_opt()` to guarantee safety.
pub struct Calendar {
    // values required for the lookup of the years/months, considering leap Februaries
    // - year_offset_ms, taking into account leap/non leap years. store in array with implied index starting at 1970
    // - month_offset_ms, for every year, taking into account leap februaries
    year_ms_offsets:             &'static [u64],
    leap_year_month_offsets:     &'static [u64],
    non_leap_year_month_offsets: &'static [u64],
}

impl Calendar {
    /// Constructor for the calendar.
    pub fn create() -> Self {
        Self {
            year_ms_offsets: YEAR_MS_OFFSETS,
            leap_year_month_offsets: LEAP_YEAR_MONTH_OFFSETS,
            non_leap_year_month_offsets: NON_LEAP_YEAR_MONTH_OFFSETS,
        }
    }

    /// Converts a `&DateTime` to ms from epoch. Note: may panic if invalid `DateTime` specified.
    /// ```rust
    /// # use chrono_light::prelude::*;
    /// let c = Calendar::create();
    /// assert_eq!(c.to_unixtime(&DateTime {year: 2010, month: 10, day: 10, hour: 10, minute: 10, second: 10, ms: 10}), 1286705410010);
    /// ```
    pub fn to_unixtime(&self, dt: &DateTime) -> u64 {
        let year = dt.year as usize - EPOCH_YEAR;
        let year_offset = self.year_ms_offsets[year];
        let month_offset = if LEAP_YEARS.contains(&(dt.year as u16)) {
            self.leap_year_month_offsets[dt.month.checked_sub(1).expect("failed to calc month - 1") as usize]
        } else {
            self.non_leap_year_month_offsets[dt.month.checked_sub(1).expect("failed to calc month - 1") as usize]
        };
        let day_offset = dt.day.checked_sub(1).expect("failed to calc day - 1") as u64 * MS_IN_DAY;
        let hour_offset = dt.hour as u64 * MS_IN_HOUR;
        let minute_offset = dt.minute as u64 * MS_IN_MIN;
        let second_offset = dt.second as u64 * MS_IN_SEC;
        let ms_offset = dt.ms as u64;

        year_offset + month_offset + day_offset + hour_offset + minute_offset + second_offset + ms_offset
    }

    /// Converts a `&DateTime` to ms from epoch, returning `Some()` if supplied `DateTime` was valid, `None` otherwise.
    /// ```rust
    /// # use chrono_light::prelude::*;
    /// let c = Calendar::create();
    /// assert_eq!(c.to_unixtime_res(&DateTime {year: 2010, month: 10, day: 10, hour: 10, minute: 10, second: 10, ms: 10}), Ok(1286705410010));
    /// assert_eq!(c.to_unixtime_res(&DateTime {year: 2010, month:  0, day: 10, hour: 10, minute: 10, second: 10, ms: 10}), Err(ValidationError::Invalid));
    /// assert_eq!(c.to_unixtime_res(&DateTime {year: 2010, month: 10, day:  0, hour: 10, minute: 10, second: 10, ms: 10}), Err(ValidationError::Invalid));
    /// ```
    pub fn to_unixtime_res(&self, dt: &DateTime) -> Result<u64, ValidationError> {
        self.validate_datetime(dt)?;
        Ok(self.to_unixtime(dt))
    }

    /// Converts ms from epoch to `DateTime`.
    pub fn from_unixtime(&self, ts: u64) -> DateTime {
        // find year
        let mut year = CURRENT_YEAR - EPOCH_YEAR;
        if ts > self.year_ms_offsets[year] {
            while ts > self.year_ms_offsets[year+1] {
                year += 1;
            }
        } else {
            year -= 1;
            while ts < self.year_ms_offsets[year] {
                year -= 1;
            }
        }
        let year_offset = ts - self.year_ms_offsets[year];
        let month_offsets = if LEAP_YEARS.contains(&(year as u16 + EPOCH_YEAR as u16)) {
            &self.leap_year_month_offsets
        } else {
            &self.non_leap_year_month_offsets
        };
        
        let mut month = 1_usize;
        if year_offset > 0 {
            while year_offset > month_offsets[month-1] {
                month += 1;
            }
            month -= 1;
        }

        let day_offset = year_offset - month_offsets[month-1];
        let day = day_offset / MS_IN_DAY + 1;
        let hour = (day_offset % MS_IN_DAY) / MS_IN_HOUR;
        let minute = (day_offset % MS_IN_HOUR) / MS_IN_MIN;
        let second = (day_offset % MS_IN_MIN) / MS_IN_SEC;
        let ms = day_offset % MS_IN_SEC;

        DateTime {
            year: (year + EPOCH_YEAR) as u16,
            month: month as u8,
            day: day as u8,
            hour: hour as u8,
            minute: minute as u8,
            second: second as u8,
            ms: ms as u16
        }
    }

    /// Given a `now` `DateTime` and `Schedule`, finds ms delta when the next occurrence should trigger.
    /// If cut of by `Schedule.end`, returns a `None`.
    pub fn next_occurrence_ms(&self, now: &DateTime, schedule: &Schedule) -> Option<u64> /* delta_in_ms */ {
        let now_in_ms = self.to_unixtime(now);
        let start_in_ms = self.to_unixtime(&schedule.start);
        let is_expired = || schedule.end.as_ref().map_or(false, |end_dt| now_in_ms > self.to_unixtime(end_dt));

        if now_in_ms < start_in_ms {
            Some(start_in_ms - now_in_ms)
        } else if is_expired() {
            None
        } else {
            let next_trigger = schedule.items.iter().map(|(freq, multiplier)| {
                match freq {
                    Frequency::Year => {
                        let m_delta = now.month as i64 - schedule.start.month as i64 + i64::from(now.to_day_unixtime() >= schedule.start.to_day_unixtime());
                        let y_delta = now.year as i64 - schedule.start.year as i64 + i64::from(m_delta > 0);
                        let total_y_from_start = ceil_div(y_delta as u32, *multiplier) * multiplier;
                        let next_occurrence = DateTime {
                            year: schedule.start.year + total_y_from_start as u16,
                            month: schedule.start.month,
                            day: schedule.start.day,
                            hour: schedule.start.hour,
                            minute: schedule.start.minute,
                            second: schedule.start.second,
                            ms: schedule.start.ms,
                        };
                        self.to_unixtime(&next_occurrence) - self.to_unixtime(now)
                    }
                    Frequency::Month => {
                        let m_delta = now.month as i64 - schedule.start.month as i64 + i64::from(now.to_day_unixtime() >= schedule.start.to_day_unixtime());
                        let total_m_delta = (now.year as i64 - schedule.start.year as i64) * 12 + m_delta;
                        let total_m_from_start = ceil_div(total_m_delta as u32, *multiplier) * multiplier;
                        let next_occurrence = DateTime {
                            year: schedule.start.year + ((schedule.start.month as u32 + total_m_from_start - 1)/12) as u16,
                            month: ((schedule.start.month as u32 + total_m_from_start - 1) % 12) as u8 + 1,
                            day: schedule.start.day,
                            hour: schedule.start.hour,
                            minute: schedule.start.minute,
                            second: schedule.start.second,
                            ms: schedule.start.ms,
                        };
                        self.to_unixtime(&next_occurrence) - self.to_unixtime(now)
                    }
                    Frequency::Week | Frequency::Day | Frequency::Hour | Frequency::Minute | Frequency::Second | Frequency::Ms => {
                        let freq_in_ms = freq.to_ms() as u64 * *multiplier as u64;
                        let ms_in_this_period = (now_in_ms - start_in_ms) % freq_in_ms;
                        if ms_in_this_period == 0 {
                            freq_in_ms
                        } else {
                            freq_in_ms - ms_in_this_period
                        }
                    },
                }
            }).min();
            // ensure trigger doesn't exceed end
            match next_trigger {
                Some(trigger) =>
                    match schedule.end.as_ref() {
                        None => Some(trigger),
                        Some(end) if now_in_ms + trigger <= self.to_unixtime(end) => Some(trigger),
                        _ => None
                    },
                _ => None
            }
        }
    }

    pub fn next_occurrence_ms_with_past_triggers(&self, last_run: Option<&DateTime>, now: &DateTime, schedule: &Schedule) -> (Vec<u64>, Option<u64>) /* triggers_in_ms, delta_in_ms */ {
        let t0 = DateTime { year: 1970, month: 1, day: 1, hour: 0, minute: 0, second: 0, ms: 0 };
        let now_ms = self.to_unixtime(&now);
        let mut triggers = Vec::new();
        let mut next_trigger: Option<u64> = None;
        let mut last_run = last_run.map(|x| x.clone()).unwrap_or(t0);
        let mut last_run_ms = self.to_unixtime(&last_run);
        while last_run_ms <= now_ms {
            next_trigger = self.next_occurrence_ms(&last_run, schedule).map(|x| x + last_run_ms);
            if let Some(trigger_ms) = next_trigger {
                let next_trigger = self.from_unixtime(trigger_ms);
                last_run = next_trigger;
                last_run_ms = trigger_ms;
                if trigger_ms <= now_ms {
                    triggers.push(trigger_ms);
                } else {
                    break;
                }
            } else {
                break;
            }
        }
        let next_trigger_delay = next_trigger.map(|x| x - now_ms);
        (triggers, next_trigger_delay)
    }

    /// Finds ms delta between 2 `DateTime`s.
    pub fn ms_between(&self, from: &DateTime, to: &DateTime) -> i64 {
        (self.to_unixtime(to) as i64).checked_sub(self.to_unixtime(from) as i64).expect("failed to calc ms_between")
    }

    /// Validates `DateTime` for correctness of fields, checking in respect to leap years.
    pub fn validate_datetime(&self, dt: &DateTime) -> Result<(), ValidationError> {
        // scope check
        (EPOCH_YEAR..=EPOCH_YEAR+self.year_ms_offsets.len()-1).contains(&(dt.year as usize));
        if !(EPOCH_YEAR..=EPOCH_YEAR+self.year_ms_offsets.len()-1).contains(&(dt.year as usize)) {
            return Err(ValidationError::OutOfScope);
        }

        // static valid check
        if !(1..=12).contains(&dt.month) || !(1..=31).contains(&dt.day) || dt.hour >= 24 || dt.minute >= 60 || dt.second >= 60 || dt.ms >= 1000 {
            return Err(ValidationError::Invalid);
        }

        // leap year check
        let is_leap_year = LEAP_YEARS.contains(&(dt.year as u16));
        if (is_leap_year && dt.day > MONTH_FOR_LEAP_YEAR[dt.month.checked_sub(1).expect("failed to calc month - 1") as usize]) ||
            (!is_leap_year && dt.day > MONTH_FOR_NON_LEAP_YEAR[dt.month.checked_sub(1).expect("failed to calc month - 1") as usize]) {
            return Err(ValidationError::Invalid);
        }
        Ok(())
    }

    pub fn validate_schedule(&self, schedule: &Schedule) -> Result<(), ValidationError> {
        self.validate_datetime(&schedule.start)?;
        if let Some(end) = &schedule.end {
            self.validate_datetime(&end)?;
        }

        let start_before_end = schedule.end.as_ref().map_or(true, |end| schedule.start <= *end);
        let all_freqs_non_zero_multiplier = schedule.items.iter().all(|&(_, x)| x > 0);
        if !start_before_end || !all_freqs_non_zero_multiplier {
            return Err(ValidationError::Invalid);
        }
        Ok(())
    }
}
